### PR TITLE
Refactor mouse dispatcher tests - Pass 1

### PR DESCRIPTION
**Pull request by aicioara**
Part of #2628

I will be doing 2 passes because I will get more context (and better sense of best practices) by refactoring other files. However, please flag anything that seems unusual.

As part of this: 
- **Test semantics: hierarchy, naming, testing, but no coverage check**
- Applied best practices from the wiki page
- Factored out common code into `beforeEach()` clauses
- No nested `beforeEach()`es
- Cleaned the console
- Remade hierarchy as shown bellow
screen shot 2015-09-11 at 5 33 40 pm

### CODE REVIEW

File: test/dispatchers/mouseDispatcherTests.ts [hunk 1]
```diff
@@ -3,174 +3,34 @@
...
-      target.remove();
+    describe("Basic usage", () => {
+      let svg: d3.Selection<void>;
+
+      beforeEach(() => {
+        svg = TestMethods.generateSVG();
+      });
+
+      it("creates only one Dispatcher.Mouse per <svg> using getDispatcher() ", () => {
+        let dispatcher1 = Plottable.Dispatchers.Mouse.getDispatcher(<SVGElement> svg.node());
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: can you factor out svg.node into a separate variable?

File: test/dispatchers/mouseDispatcherTests.ts [hunk 2]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+      });
 
-      TestMethods.triggerFakeMouseEvent("mousedown", target, targetX, targetY);
-      assert.isTrue(callbackWasCalled, "callback was called on mousedown");
+      it("calls the mouseDown callback", () => {
+        let callbackWasCalled = false;
+        let callback = (point: Plottable.Point, event: MouseEvent) => {
+          callbackWasCalled = true;
+          TestMethods.assertPointsClose(point, expectedPoint, 0.5, "mouse position is correct");
+          assert.isNotNull(event, "mouse event was passed to the callback");
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: also check if event is an instance of MouseEvent?

File: test/dispatchers/mouseDispatcherTests.ts [hunk 3]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+        }
+
+        let targetDeltaY = 10;
+
+        let callbackWasCalled = false;
+        let callback = (point: Plottable.Point, event: WheelEvent) => {
+          callbackWasCalled = true;
+          assert.strictEqual(event.deltaY, targetDeltaY, "deltaY value was passed to callback");
+          TestMethods.assertPointsClose(point, expectedPoint, 0.5, "mouse position is correct");
+          assert.isNotNull(event, "mouse event was passed to the callback");
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: same here

File: test/dispatchers/mouseDispatcherTests.ts [hunk 4]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+
+        svg.remove();
+      });
+
+      it("calls the dblClick callback", () => {
+        let callbackWasCalled = false;
+        let callback = (point: Plottable.Point, event: MouseEvent) => {
+          callbackWasCalled = true;
+          TestMethods.assertPointsClose(point, expectedPoint, 0.5, "mouse position is correct");
+          assert.isNotNull(event, "mouse event was passed to the callback");
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: check event type

File: test/dispatchers/mouseDispatcherTests.ts [hunk 5]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+        assert.isTrue(cb2Called, "callback 2 was called on mousemove");
+
+        cb1Called = false;
+        cb2Called = false;
+        mouseDispatcher.offMouseMove(cb1);
+        TestMethods.triggerFakeMouseEvent("mousemove", svg, targetX, targetY);
+        assert.isFalse(cb1Called, "callback was not called after blanking");
+        assert.isTrue(cb2Called, "callback 2 was still called");
+
+        svg.remove();
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: unregister cb2 as well

File: test/dispatchers/mouseDispatcherTests.ts [hunk 6]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+        let callbackWasCalled = false;
+        let callback = () => callbackWasCalled = true;
+
+        mouseDispatcher.onMouseDown(callback);
+        TestMethods.triggerFakeMouseEvent("mousedown", svg, targetX, targetY);
+        assert.isTrue(callbackWasCalled, "callback was called on mousedown");
+
+        let element = <HTMLElement> svg[0][0];
+        let position = { x: 0, y: 0 };
+        while (element != null) {
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: a comment about this?

File: test/dispatchers/mouseDispatcherTests.ts [hunk 7]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+        callbackWasCalled = false;
+        TestMethods.triggerFakeMouseEvent("mousedown", overlay, targetX, targetY);
+        assert.isFalse(callbackWasCalled, "callback was not called on mousedown on overlay");
+
+        mouseDispatcher.offMouseDown(callback);
+        svg.remove();
+        overlay.remove();
+      });
+
+      it("calls callbacks on mouseover, mousemove, and mouseout", () => {
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: calls mousemove callback  on...

File: test/dispatchers/mouseDispatcherTests.ts [hunk 8]
```diff
@@ -178,117 +38,219 @@ describe("Dispatchers", () => {
...
+        TestMethods.triggerFakeMouseEvent("mouseover", svg, targetX, targetY);
+        assert.isTrue(callbackWasCalled, "callback was called on mouseover");
+        callbackWasCalled = false;
+        TestMethods.triggerFakeMouseEvent("mousemove", svg, targetX, targetY);
+        assert.isTrue(callbackWasCalled, "callback was called on mousemove");
+        callbackWasCalled = false;
+        TestMethods.triggerFakeMouseEvent("mouseout", svg, targetX, targetY);
+        assert.isTrue(callbackWasCalled, "callback was called on mouseout");
+
+        mouseDispatcher.offMouseMove(callback);
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: check it doesn't call callbacks after disconnecting?

File: test/dispatchers/mouseDispatcherTests.ts (hunk 2 above)
Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: We don't do this anywhere else though

Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: we probably didn't but I think it's good for completeness
personally I would prefer if you can test if it's the same `event` that you created via `TestMethods.triggerFakeMouseEvent`but that could be messy and requires some rewrite in `TestMethods`
so I am okay with checking if its type as bare minimum

Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: sounds good

File: test/dispatchers/mouseDispatcherTests.ts (hunk 8 above)
Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: fixed

File: test/dispatchers/mouseDispatcherTests.ts (hunk 4 above)
Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: fixed

File: test/dispatchers/mouseDispatcherTests.ts (hunk 3 above)
Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: fixed

File: test/dispatchers/mouseDispatcherTests.ts (hunk 2 above)
Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: fixed

File: test/dispatchers/mouseDispatcherTests.ts [hunk 9]
```diff
@@ -141,154 +40,235 @@ describe("Dispatchers", () => {
...
+        mouseDispatcher = Plottable.Dispatchers.Mouse.getDispatcher(<SVGElement> svg.node());
+      });
+
+      it("calls the mouseDown callback", () => {
+        let callbackWasCalled = false;
+        let callback = (point: Plottable.Point, event: MouseEvent) => {
+          callbackWasCalled = true;
+          TestMethods.assertPointsClose(point, expectedPoint, 0.5, "mouse position is correct");
+          assert.isNotNull(event, "mouse event was passed to the callback");
+          assert.isTrue(event instanceof MouseEvent, "the event passed is an instance of MouseEvent");
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: please use `assert.instanceOf`
same below

File: test/dispatchers/mouseDispatcherTests.ts [hunk 10]
```diff
@@ -141,154 +40,235 @@ describe("Dispatchers", () => {
...
+        assert.isFalse(callbackWasCalled, "disconnected dispatcher callback not called on mouseover");
+        TestMethods.triggerFakeMouseEvent("mousemove", svg, targetX, targetY);
+        assert.isFalse(callbackWasCalled, "disconnected dispatcher callback not called on mousemove");
+        TestMethods.triggerFakeMouseEvent("mouseout", svg, targetX, targetY);
+        assert.isFalse(callbackWasCalled, "disconnected dispatcher callback not called on mouseout");
+
+        svg.remove();
+      });
+
+      it("can register two callbacks for the samme mouse dispatcher", () => {
```
Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: same

Comment (aicioara) [inline test/dispatchers/mouseDispatcherTests.ts]: Fixed

Comment (ztsai) [inline test/dispatchers/mouseDispatcherTests.ts]: ?

**Pull request merged**